Menus that use data

jneem opened the pull request:
This is a draft of my attempt at #965

jneem commented:
I think this is ready for a look.

I think the new mac behavior is an improvement over what was there before, although I haven't been able to test it. Now, it keeps track of the most-recently-focused window with a menu and always treats that one as the global app menu. Previously, it would set the menu whenever the focus changed, but it would also change it on every `SET_MENU` command, meaning that a background window changing its menu would override the global app menu.

cmyr commented:
@jneem very cool, will take a look at this in the next day or two!

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! ## Menus
```
The first line of the module comments should just be a short plug, like "Window and applications menus"; it will show up as the short description beside this module on the main docs page.

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+//!
+//! Menu items can be associated with callbacks, which are triggered when a user selects that menu
+//! item. Each callback has access to the application data, and also gets access to a
+//! [`MenuEventCtx`], which allows for submitting commands.
```
I'd link ```[`Command`]```s, here?

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+
+//! ## Menus
+//!
+//! Menus in `druid` follow a data-driven design similar to that of the main widget tree. The main
```
Druid style guide: We say "Druid" when referring generally to the project, and "`druid`" only when we're talking explicitly about the crate, as in "Be sure to test both `druid` and `druid-shell` manually." This use is ambiguous, but informed by some use below I think we want 
"Druid".

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+
+impl<T: Data> ContextMenu<T> {
+    /// Create a new [`ContextMenu`].
+    pub fn new(
```
Just want to doublecheck my understanding here: I can construct my `ContextMenu` from *any* data, and not just the root app state, correct?

jneem replied:
No, currently the contexts' `show_context_menu` function still checks that we're using the root data. In principle, we could probably do something like what `new_sub_window` does, though.

cmyr replied:
So playing around with this in runebender, I wonder if we can simplify the context menu case?

My basic thinking: a context menu is a modal. It should never need to be updated, since it is dismissed if anything is clicked. Additionally, in a non-trivial application a context menu is going to be presented somewhere fairly deep in the widget hierarchy, where the root data is likely to be several lenses away.

For me, I think that having access to the root data (especially when it's basically untyped) in the context menu just isn't that helpful. I would much prefer if, in this case, I instantiate a single menu object and display it each time.

Does this make sense? It's a bit different from the API for other menus, but ultimately feels simpler than what we have now and potentially more flexible, since you can pass in whatever data you want when building your menu.

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+        ret
+    }
+
+    pub fn refresh(&mut self, data: &T, env: &Env) -> PlatformMenu {
```
docs plz

jneem replied:
Shouldn't have been pub, sorry...

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+
+/// This is the trait that enables recursive visiting of all menu entries. It isn't publically
+/// visible (the publically visible analogue of this is `Into<MenuEntry<T>>`).
+trait MenuTrait<T> {
```
Maybe just `MenuVisitor`?

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+        }
+    }
+
+    pub fn with_submenu(&mut self, text: &str, enabled: bool, f: impl FnOnce(&mut MenuBuildCtx)) {
```
It looks like all of these methods are actually internal, in which case we can remove the `pub`/`pub (crate)`, for clarity?

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+        let mut ctx = MenuBuildCtx::new(self.popup);
+        self.menu
+            .as_mut()
+            .unwrap()
```
this is purely academic, but lately I've been curious about the possible code-size costs of unwraps, and am wondering if it's worth trying to avoid them in places like this where it's relatively trivial.

This is not a suggestion, I'm just mentioning it, it's something I'd like to try measuring at some point.

jneem replied:
It's a good question. I actually think I like it better with an "if let", so I got rid of the unwraps anyway. I wouldn't be terribly surprised, though, if rustc is smart enough to remove it in cases like this (https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=9f50ee6b6740c01277b4f680658deb31)

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+    needs_refresh: Option<MenuPredicate<T>>,
+    item: MenuItem<T>,
+    children: Vec<MenuEntry<T>>,
+    // bloom?
```
This would require children to have some sort of identity, and for there to be some benefit in only visiting specific children. Do you think this would make sense?

jneem replied:
The `MenuItem`s are supposed to have unique ids, so blooms would basically allow us to traverse directly to the right `MenuItem` when getting a menu command from the platform.

I didn't do it yet because:
- the menu probably can't be big enough (or commands common enough) for traversing the menu tree to be a bottleneck, and
- it might actually be more expensive to construct bloom filters each time the menu is rebuilt (I have more to say about rebuilding, but I'll put that elsewhere later)

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+    /// supplying a refresh callback, the menu will recursively check whether any children have
+    /// changed and refresh itself if any have. By supplying a callback here, you can short-circuit
+    /// those recursive calls.
+    pub fn needs_refresh(mut self, refresh: impl FnMut(&T, &T, &Env) -> bool + 'static) -> Self {
```
I'm not in love with either of these two names, although I'm not sure I have a better suggestion. `refresh_on`? `refresh_fn`? `refresh_check`? `check_refresh`? `with_refresh_check`? I don't really like any of these, either.

jneem replied:
My current favorite (using the term a little loosely) is `refresh_on`

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+        self
+    }
+
+    /// Supply a function to check when this menu needs to be rebuild from scratch.
```
```suggestion
    /// Supply a function to check when this menu needs to be rebuilt from scratch.
```

jneem replied:
I'm just gonna do this in my local copy to avoid an extra rebase

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+/// [`MenuItem::lens`] instead of using this struct directly.
+///
+/// [`LensWrap`]: crate::widget::LensWrap
+pub struct MenuLensWrap<L, U> {
```
we could consider also folding some of the `WidgetPod` functionality into here; in particular we could stash the last seen data and env, and then stop recursing if they're unchanged?

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,762 @@
+    /// Provide a dynamic hotkey for activating this menu item.
+    ///
+    /// The hotkey can change depending on the data.
+    pub fn dynamic_hotkey(
```
I like this :)

cmyr reviewed (commented):
This looks great! I've left some preliminary comments inline, although I haven't tried playing around with it. My initial impression is that this looks like a huge improvement, though, thanks for taking it on :)

jneem commented on druid/src/app.rs:
```diff
@@ -120,8 +121,8 @@ impl<T: Data> PendingWindow<T> {
     /// Set the menu for this window.
-    pub fn menu(mut self, menu: MenuDesc<T>) -> Self {
-        self.menu = Some(menu);
+    pub fn menu(mut self, menu: impl FnMut(&T, &Env) -> Menu<T> + 'static) -> Self {
```
Should the context menus also get a `WindowId` parameter?

cmyr replied:
Hm, no I don't think so? I expect them to be more transient, and you can pass extra information in when you construct them.

jneem commented:
Thanks for your review! I think I addressed your comments except for the `context_menu` one. Next, I'm going to go spend some time dogfooding this in my own druid project.

One remark about rebuilding: this implementation internally rebuilds the entire druid-shell menu from scratch each time anything changes, but I think that as long as the underlying platform API cooperates (I haven't checked anything except for GTK), we could do the updates entry-by-entry. It's a bit of a pain trying to coordinate the druid-shell changes across backends, so I haven't tried to do it here.

cmyr commented:
Okay I'll take a closer look at the context menu situation and play around with it a bit; context menus are often presented in situations where the user doesn't have access to the root data (like in some widget deep in the tree) so I want to make sure there's a reasonable solution, there.

cmyr commented on druid/src/menu/mod.rs:
```diff
@@ -0,0 +1,790 @@
+    }
+
+    /// Append a menu entry to this menu, returning the modified menu.
+    pub fn with_entry(mut self, entry: impl Into<MenuEntry<T>>) -> Self {
```
It doesn't *really* matter, but I might consider including the old method names as aliases, like:

```rust
#[doc(hidden)]
#[deprecated(since = "0.8.0", note = "use with_entry instead")]
pub fn append(mut self, entry: impl Into<MenuEntry<T>>) -> Self {
    self.with_entry(entry)
}
```

I might also consider having,

```rust
#[deprecated(since = "0.8.0", note = "Renamed to Menu")]
pub type MenuDesc<T> = Menu<T>;
```

anyway I'm sort of anal-retentive about this sort of stuff, the breakage isn't a big deal.

cmyr reviewed (changes requested):
Okay so I updated runebender to use this and ran into a few problems; the major one is that this breaks the built-in commands for open files etc.

There are a few other thoughts inline; in any case this is exciting. :)

jneem commented on druid/src/win_handler.rs:
```diff
@@ -174,14 +174,20 @@ impl<T> AppState<T> {
-        }
+                .as_mut()
+                .map(|m| m.event(queue, cmd_id, data, env)),
+        };
```
Thanks for the pointer! I ran into this myself today, but hadn't tracked down the reason yet...

jneem replied:
Ok, so having dug into this a bit more, I'm not sure I understand your comment. `m.event` is the event method on `MenuManager`, which is basically responsible for pushing commands on the command queue, and then those commands get dispatched using the usual `AppState` command dispatching/updating mechanism. I did have a bug, however, where the target wasn't being set, which caused some commands to get dropped...

cmyr replied:
oooh, so here the menu receives the `cmd_id`, and then submits the command via the context? That sounds good. 👍

jneem commented:
Dunno why, but github wouldn't let me reply to this directly:

> So playing around with this in runebender, I wonder if we can simplify the context menu case?

> My basic thinking: a context menu is a modal. It should never need to be updated, since it is dismissed if anything is clicked. Additionally, in a non-trivial application a context menu is going to be presented somewhere fairly deep in the widget hierarchy, where the root data is likely to be several lenses away.

> For me, I think that having access to the root data (especially when it's basically untyped) in the context menu just isn't that helpful. I would much prefer if, in this case, I instantiate a single menu object and display it each time.

> Does this make sense? It's a bit different from the API for other menus, but ultimately feels simpler than what we have now and potentially more flexible, since you can pass in whatever data you want when building your menu.

Are you just talking about the menu construction here, or more generally about the various data-accessing callbacks in the menu? I definitely agree that we don't need the "builder" callback. I also think that things like `enabled` taking a data callback are overkill, but I do think I'd like to have access to the (locally lensed, ideally) data in the `on_activated` callback. (One of my motivations for doing this rewrite was that I got annoyed at how menus forced me to do everything through commands.) What would you think about this API change:

- make the context menu constructor take just the menu, not a "builder" function
- make the context menu use the local data, the same way that subwindows do (I haven't thought through how this would be implemented, though)
- add convenience functions for non-dynamic properties, like the way we have both `hotkey` and `dynamic_hotkey`

cmyr commented:
Those ideas make sense, I think? I'm not sure about the local data bit; it isn't totally obvious to me how that would work. What I could imagine, although it's a bit less ergonomic, is making the caller build the menu themselves, using the available data and env, and passing the `PlatformMenu` to `show_context_menu`?

I'm generally a fan ofconvenience methods.

jneem commented:
Ok, I think I've addressed the issues except for the question about access to the local data in context menus (which I haven't even looked into, and I'd prefer to punt on for now). What I did instead was make it so that all the relevant menu-building functions have both dynamic and static versions, so it's now much more convenient to construct a menu based on the current data that doesn't need access to future data.

cmyr commented:
Will take another swing at this soon!

cmyr commented:
I really like the API tweaks here, but I do have one lingering problem with the `ContextMenu`. Is there some way to let me create a `ContextMenu` with just a single, concrete `Menu` object? It's really a bit of a brain-twister that the function I use to create a menu is passed a type I don't have easy access to, and can not easily extract the information I need from, because the path from there to here involves passing through multiple non-trivial lenses.

I'm not trying to be difficult, here; I legitimately spent 20+ minutes getting the context menu to work in runebender, and it ended up with me just cloning my local data and moving that into the closure, ignoring the closure arguments altogether, like:

```rust
            let data = data.clone();
            let menu: ContextMenu<AppState> = ContextMenu::new(
                move |_, _| crate::menus::make_context_menu(&data, mouse_pos),
                m.window_pos,
            );
```

and this doesn't feel like a great idiom.

Other than that, this looks good. The window menus are working now and I'm not seeing the problem with dropped commands, although when I'm slightly more alert I will go and review my last comment there to make sure I'm not totally losing it.

Thanks again, this feels great!

jneem commented:
Sorry, with all the other API changes that were supposed to make context menus more convenient, the one you actually asked for slipped through the cracks...

cmyr commented:
Great that looks like a solid simplification.

Also I'm seeing a warning from CI about an 'unread field'?

I'm going to take one more pass through the root command handling logic, but modulo that this looks good to go!

cmyr commented on druid/src/win_handler.rs:
```diff
@@ -93,7 +93,10 @@ struct Inner<T> {
+    root_menu: Option<MenuManager<T>>,
+    /// The id of the most-recently-focused window that has a menu. On macOS, this
+    /// is the window that's currently in charge of the app menu.
+    menu_window: Option<WindowId>,
```
ah okay I'm seeing an 'unused field' warning on this for non-mac platforms. Maybe worth an 'allow unused' or whatever?

jneem replied:
Strange that it doesn't complain for me locally... maybe it's a different behavior in nightly? :shrug: Anyway, easy to fix..

cmyr reviewed (approved):
Looks good, thanks again for taking this on, it's a solid improvement.
